jbpm-flow/src/main/java/org/jbpm/workflow/instance/impl/NodeInstanceImpl.java

```diff
@@ -257,113 +264,119 @@ protected void triggerCompleted(String type, boolean remove) {
-    				if (groupInstance.containsNodeInstance(this)) {
-    					for (NodeInstance nodeInstance: groupInstance.getNodeInstances()) {
-    						if (nodeInstance != this) {
-    							((org.jbpm.workflow.instance.NodeInstance) nodeInstance).cancel();
-    						}
-    					}
-    					((ContextInstanceContainer) parent).removeContextInstance(ExclusiveGroup.EXCLUSIVE_GROUP, contextInstance);
-    				}
-    				
-    			}
-    		}
-    	}
-    	return (org.jbpm.workflow.instance.NodeInstance)
-    		((org.jbpm.workflow.instance.NodeInstanceContainer) getNodeInstanceContainer())
-            	.getNodeInstance(connection.getTo());
+        // check for exclusive group first
+        NodeInstanceContainer parent = getNodeInstanceContainer();
+        if (parent instanceof ContextInstanceContainer) {
+            List<ContextInstance> contextInstances = ((ContextInstanceContainer) parent).getContextInstances(ExclusiveGroup.EXCLUSIVE_GROUP);
+            if (contextInstances != null) {
+                for (ContextInstance contextInstance: new ArrayList<ContextInstance>(contextInstances)) {
+                    ExclusiveGroupInstance groupInstance = (ExclusiveGroupInstance) contextInstance;
+                    if (groupInstance.containsNodeInstance(this)) {
+                        for (NodeInstance nodeInstance: groupInstance.getNodeInstances()) {
+                            if (nodeInstance != this) {
+ ((org.jbpm.workflow.instance.NodeInstance) nodeInstance).cancel();
+                            }
+                        }
+                        ((ContextInstanceContainer) parent).removeContextInstance(ExclusiveGroup.EXCLUSIVE_GROUP, contextInstance);
+                    }
+
+                }
+            }
+        }
+        return (org.jbpm.workflow.instance.NodeInstance)
+            ((org.jbpm.workflow.instance.NodeInstanceContainer) getNodeInstanceContainer())
+                .createNodeInstance(connection.getTo());
     }
-    
+
+    // NOTES NodeInstanceImpl: end of mini-stack
     protected void triggerNodeInstance(org.jbpm.workflow.instance.NodeInstance nodeInstance, String type) {
-    	boolean hidden = false;
-    	if (getNode().getMetaData().get("hidden") != null) {
-    		hidden = true;
-    	}
-    	InternalKnowledgeRuntime kruntime = getProcessInstance().getKnowledgeRuntime();
-    	if (!hidden) {
-    		((InternalProcessRuntime) kruntime.getProcessRuntime())
-    			.getProcessEventSupport().fireBeforeNodeLeft(this, kruntime);
-    	}
-    	// trigger next node
-        nodeInstance.trigger(this, type);
+        boolean hidden = isHidden();
+        InternalKnowledgeRuntime kruntime = getProcessInstance().getKnowledgeRuntime();
+        if (!hidden) {
+            ((InternalProcessRuntime) kruntime.getProcessRuntime())
+                .getProcessEventSupport().fireBeforeNodeLeft(this, kruntime);
+        }
+        if( isStackless() ) {
+            addNodeInstanceTrigger(nodeInstance, this, type);
```

Comment:
I'm not sure, although I understand what you mean. I'll look at trying to make it more elegant. 

However, what you're suggesting would make the code less readable and would decrease performance for the old execution model, which is why I'm reluctant to do it.